### PR TITLE
executor: refactor merge join

### PR DESCRIPTION
**Pull request by zz-jason**
1. split `readerIterator` to `mergeJoinInnerTable` and `mergeJoinOuterTable`, use `mergeJoinInnerTable` to returns all the inner rows within the same join group, use `mergeJoinOuterTable` to return a Chunk of outer rows and iterate on it when we do the merge join process.
`(rather than buffer all the outer rows with the same join key, which can also reduce the memory usage during merge join)`

2. fix a potential bug caused by always appending the unmatched outer row to the very beginning output Chunk. `(caused by readerIterator.nextSelectedRow() )`,
this can also promise the result order according to the output rows.

### CODE REVIEW

**Comment by zz-jason** [comment]
```sql
drop table if exists t1, t2;
create table t1(a bigint primary key, b bigint);
create table t2(a bigint primary key, b bigint);
insert into t1 values(1, 100), (2, 100), (3, 100), (4, 100), (5, 100);
insert into t2 select a*100, b*100 from t1;
```

before this PR:
```sql
set @@tidb_max_chunk_size=1;

TiDB(localhost) > select /*+ TIDB_SMJ(t2) */ * from t1 left outer join t2 on t1.a=t2.a and t1.a!=3 order by t1.a;
+---+------+---+------+
| a | b    | a | b    |
+---+------+---+------+
| 1 |  100 | NULL | NULL |
| 3 |  100 | NULL | NULL |
| 2 |  100 | NULL | NULL |
| 4 |  100 | NULL | NULL |
| 5 |  100 | NULL | NULL |
+---+------+---+------+
5 rows in set (0.00 sec)
```

after this PR:
```sql
set @@tidb_max_chunk_size=1;

TiDB(localhost) > select /*+ TIDB_SMJ(t2) */ * from t1 left outer join t2 on t1.a=t2.a and t1.a!=3 order by t1.a;select /*+ TIDB_SMJ(t2) */ * from t1 left outer join t2 on t1.a=t2.a and t1.a!=3 order by t1.a;
+---+------+---+------+
| a | b    | a | b    |
+---+------+---+------+
| 1 |  100 | NULL | NULL |
| 2 |  100 | NULL | NULL |
| 3 |  100 | NULL | NULL |
| 4 |  100 | NULL | NULL |
| 5 |  100 | NULL | NULL |
+---+------+---+------+
5 rows in set (0.00 sec)
```

**Comment by zz-jason** [comment]
@XuHuaiyu @winoros PTAL

File: executor/merge_join.go [hunk 1]
```diff
@@ -249,72 +221,69 @@ func (e *MergeJoinExec) NextChunk(ctx context.Context, chk *chunk.Chunk) error {
...
-			if e.innerIter4Row.Current() == e.innerIter4Row.End() {
-				e.outerIter4Row.Next()
-				e.innerIter4Row.Begin()
-			}
-			if chk.NumRows() == e.maxChunkSize {
-				return true, errors.Trace(err)
+
+			e.outerTable.row = e.outerTable.iter.Next()
+
+			if chk.NumRows() >= e.maxChunkSize {
```
Comment (XuHuaiyu) [inline executor/merge_join.go]: s/ >=/ ==
since chk.NumRows can be promised not greater than maxChunkSize.

Comment (zz-jason) [reply executor/merge_join.go]: I think `>=` is more safe.

**Comment by XuHuaiyu** [comment]
PTAL @winoros

**Comment by shenli** [comment]
@XuHuaiyu @winoros PTAL

File: executor/merge_join.go [hunk 2]
```diff
@@ -33,27 +32,41 @@ type MergeJoinExec struct {
...
+	t.chk = chk
+	t.iter = chunk.NewIterator4Chunk(chk)
+	t.selected = make([]bool, 0, maxChunkSize)
 }
 
-// readerIterator represents a row block with the same join keys
-type readerIterator struct {
-	stmtCtx  *stmtctx.StatementContext
-	sctx     sessionctx.Context
+// mergeJoinInnerTable represents inner table for merge join, which returns a
```
Comment (XuHuaiyu) [inline executor/merge_join.go]: mergeJoinInnerTable indicates inner table of merge join, which returns a
batch of rows with the same join key.

File: executor/merge_join.go [hunk 3]
```diff
@@ -323,12 +302,20 @@ func (e *MergeJoinExec) fetchNextInnerRows() (err error) {
 	return nil
 }
 
-func (e *MergeJoinExec) fetchNextOuterRows() (err error) {
-	e.outerRows, err = e.outerIter.rowsWithSameKey()
+// fetchNextOuterRows fetches the next Chunk of outer table. Rows in a Chunk
```
Comment (XuHuaiyu) [inline executor/merge_join.go]: fetchNextOuterRows fetches the Next Chunk from outer table.
All the rows in this Chunk are not promised to have the same join key,
but promised to be sorted according to the join keys.

File: executor/merge_join.go [hunk 4]
```diff
@@ -249,72 +229,71 @@ func (e *MergeJoinExec) NextChunk(ctx context.Context, chk *chunk.Chunk) error {
...
 		}
 
+		if chk.NumRows() >= e.maxChunkSize {
+			return true, errors.Trace(err)
+		}
 	}
-	return false, nil
 }
 
+// fetchNextInnerRows fetches the next join group belongs to the same join key
```
Comment (XuHuaiyu) [inline executor/merge_join.go]: s/ belongs to/ with

File: executor/merge_join.go [hunk 5]
```diff
@@ -249,72 +229,71 @@ func (e *MergeJoinExec) NextChunk(ctx context.Context, chk *chunk.Chunk) error {
...
+// on inner table.
```
Comment (XuHuaiyu) [inline executor/merge_join.go]: s/ on/ from

File: executor/builder.go [hunk 6]
```diff
@@ -657,52 +674,38 @@ func (b *executorBuilder) buildMergeJoin(v *plan.PhysicalMergeJoin) Executor {
...
+		innerFilter = v.LeftConditions
+		e.innerTable.reader = leftExec
+		e.innerTable.joinKeys = leftKeys
 	}
 
-	if v.JoinType != plan.InnerJoin {
-		e.outerFilter = e.outerIter.filter
-		e.outerIter.filter = nil
+	// optimizer should guarantee that filters on inner table are pushed down
+	// to tikv or extracted to a Selection.
```
Comment (XuHuaiyu) [inline executor/builder.go]: s/ extracted/ be extracted

File: executor/merge_join.go [hunk 7]
```diff
@@ -33,27 +32,41 @@ type MergeJoinExec struct {
...
+	keys   []*expression.Column
+
+	chk      *chunk.Chunk
+	selected []bool
+
+	iter *chunk.Iterator4Chunk
+	row  chunk.Row
+}
+
+func (t *mergeJoinOuterTable) init(chk *chunk.Chunk, maxChunkSize int) {
```
Comment (XuHuaiyu) [inline executor/merge_join.go]: remove this func

File: executor/merge_join.go [hunk 8]
```diff
@@ -33,27 +32,36 @@ type MergeJoinExec struct {
...
+	resultGenerator joinResultGenerator
+
+	prepared bool
+	outerIdx int
+
+	innerTable *mergeJoinInnerTable
+	outerTable *mergeJoinOuterTable
+
+	innerRows     []chunk.Row
+	innerIter4Row chunk.Iterator
```
Comment (coocood) [inline executor/merge_join.go]: How about `innerSameKeyRows` and `innerSameKeyIter`?
I can be easier to read.

**Pull request merged**